### PR TITLE
feat: add mypy plugin

### PR DESCRIPTION
**Pull request by hiro-o918**
## What
* implement mypy plugin for `gokart.TaskOnKart`
  *  This plugin checks the following when all the parameters are typed:
    * The parameters passed by the constructor are valid type
    * There are no missing parameters

image

### CODE REVIEW

File: gokart/testing/check_if_run_with_empty_data_frame.py [hunk 1]
```diff
@@ -81,7 +81,8 @@ def _test_run_with_empty_data_frame(cmdline_args: List[str], test_run_params: te
 
 def try_to_run_test_for_empty_data_frame(cmdline_args: List[str]):
     with CmdlineParser.global_instance(cmdline_args):
-        test_run_params = test_run()
+        # NOTE: parameters passed by command line cannot be linted by mypy.
+        test_run_params = test_run()  # type: ignore
```
Comment (hiro-o918) [inline gokart/testing/check_if_run_with_empty_data_frame.py:85]: mypy-plugin cannot pass this case even though the parameters passed from `cmdlin_args`

File: gokart/mypy.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+METADATA_TAG: Final[str] = 'task_on_kart'
+
+
+class TaskOnKartPlugin(Plugin):
+    def get_base_class_hook(self, fullname: str) -> Callable[[ClassDefContext], None] | None:
+        if 'gokart.task.luigi.Task' in fullname:
+            # gather attibutes from gokart.TaskOnKart
+            # the transformation does not affect because the class has `__init__` method
+            return self._task_on_kart_class_maker_callback
```
Comment (hiro-o918) [inline gokart/mypy.py:68]: This is an entrypoint to transform a class to have a `__init__` method with `luigi. Parameter`s while type-checking for `gokart.TaskOnKart` class

File: gokart/mypy.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+        sym = self.lookup_fully_qualified(fullname)
+        if sym and isinstance(sym.node, TypeInfo):
+            if any(base.fullname == 'gokart.task.TaskOnKart' for base in sym.node.mro):
```
Comment (hiro-o918) [inline gokart/mypy.py:72]: This is also an entrypoint for inheritance of `gokart.TaskOnKart`

File: gokart/mypy.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            if attr.type is None:
+                return False
+
+        # If there are no attributes, it may be that the semantic analyzer has not
+        # processed them yet. In order to work around this, we can simply skip generating
+        # __init__ if there are no attributes, because if the user truly did not define any,
+        # then the object default __init__ with an empty signature will be present anyway.
+        if ('__init__' not in info.names or info.names['__init__'].plugin_generated) and attributes:
+            args = [attr.to_argument(info, of='__init__') for attr in attributes]
+            add_method_to_class(self._api, self._cls, '__init__', args=args, return_type=NoneType())
```
Comment (hiro-o918) [inline gokart/mypy.py:180]: This is a point which create a method `__init__` to a class while type checking

File: test/test_mypy.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    bar: str = luigi.Parameter() # type: ignore
+
+MyTask(foo='1') # foo is int, but it is str and bar is missing.
+        """
+
+        with tempfile.NamedTemporaryFile(suffix='.py') as test_file:
+            test_file.write(test_code.encode('utf-8'))
+            test_file.flush()
+            result = api.run(['--config-file', str(PYPROJECT_TOML), test_file.name])
+            self.assertIn('error: Missing named argument "bar" for "MyTask"  [call-arg]', result[0])
```
Comment (hiro-o918) [inline test/test_mypy.py:50]: These are error messages for invalid arguments

File: pyproject.toml [hunk 6]
```diff
@@ -64,6 +64,7 @@ quote-style = "single"
 
 [tool.mypy]
 ignore_missing_imports = true
+plugins = ["gokart.mypy:plugin"]
```
Comment (hirosassa) [inline pyproject.toml:67]: 👍 👍 👍 👍 👍 👍

**Review by hirosassa — APPROVED** [review]
GREAT PR! Thanks!

**Comment by yamamuteki** [comment]
👍

**Review by kitagry — CHANGES_REQUESTED** [review]
I will review it thoroughly today.

File: gokart/mypy.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+            yield from self._get_assignment_statements_from_block(stmt.else_body)
+
+    def _get_assignment_statements_from_block(self, block: Block) -> Iterator[AssignmentStmt]:
+        for stmt in block.body:
+            if isinstance(stmt, AssignmentStmt):
+                yield stmt
+            elif isinstance(stmt, IfStmt):
+                yield from self._get_assignment_statements_from_if_statement(stmt)
+
+    def collect_attributes(self) -> list[TaskOnKartAttribute] | None:
```
[the remark this answers is not in the document]
Comment (hiro-o918) [reply gokart/mypy.py]: [e735a48](https://github.com/m3dev/gokart/pull/384/commits/e735a48b2763cf399b69f835bc54ac075e7c88dd)

File: docs/mypy_plugin.rst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+.. code:: python
+
+    import gokart
+    import luigi
+
+
+    class Foo(gokart.TaskOnKart):
+        # NOTE: must all the parameters be annotated
+        foo: int = luigi.IntParameter(default=1)
+        bar: str = luigi.IntParameter()
```
Comment (kitagry) [inline docs/mypy_plugin.rst:37]:
```suggestion
        bar: str = luigi.Parameter()
```

File: test/test_mypy.py (hunk 5 above)
Comment (kitagry) [inline test/test_mypy.py:50]: Is it possible to ignore missing named argument error? In some cases, we set the parameter with config file.

https://gokart.readthedocs.io/en/latest/setting_task_parameters.html#set-parameter-at-config-file

File: gokart/mypy.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+        t = get_proper_type(sym.type)
+
+        # Perform a simple-minded inference from the signature of __set__, if present.
+        # We can't use mypy.checkmember here, since this plugin runs before type checking.
+        # We only support some basic scanerios here, which is hopefully sufficient for
+        # the vast majority of use cases.
+        if not isinstance(t, Instance):
+            return default
+        setter = t.type.get('__set__')
+        if setter:
```
Comment (kitagry) [inline gokart/mypy.py:340]: it seems to be prefer to return early.

```python
if not setter:
    return default
```

**Review by kitagry — CHANGES_REQUESTED** [review]

File: test/test_mypy.py (hunk 5 above)
Comment (hiro-o918) [reply test/test_mypy.py:50]: Ok, I'll check this

Comment (hiro-o918) [reply test/test_mypy.py]: made this optional.
We can make the parameters optional or required based on reading yaml, which is future work.

[08e2b74](https://github.com/m3dev/gokart/pull/384/commits/08e2b74a4b96dec5a2f5c268962b405a2ce54b18)

File: gokart/mypy.py (hunk 9 above)
Comment (hiro-o918) [reply gokart/mypy.py]: [83e1f53](https://github.com/m3dev/gokart/pull/384/commits/83e1f531831ad119f6ea0a8cfa9c7bfb140821cc)

File: docs/mypy_plugin.rst (hunk 8 above)
Comment (hiro-o918) [reply docs/mypy_plugin.rst]: [96ea087](https://github.com/m3dev/gokart/pull/384/commits/96ea087ca26c425d83f6d61c434db4daad0ba4fe)

**Review by kitagry — APPROVED** [review]
LGTM!

**Pull request merged**